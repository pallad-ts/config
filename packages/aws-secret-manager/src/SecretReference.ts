export interface SecretReference {
	name: string;
	property?: string;
}

export namespace SecretReference {
	export function computeDescription({ name, property }: SecretReference) {
		return `AWS Secret Manager secret name: ${name}${property ? `, property: ${property}` : ""}`;
	}
}
