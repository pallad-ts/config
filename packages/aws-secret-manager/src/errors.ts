import { Domain, ErrorDescriptor, formatCodeFactory } from "@pallad/errors";

export const errorsDomain = new Domain();

const code = formatCodeFactory("E_CONF_SECRET_MANAGER_%c");
export const ERRORS = errorsDomain.addErrorsDescriptorsMap({
	CANNOT_EXTRACT_PROPERTY_FROM_NON_OBJECT: ErrorDescriptor.useDefaultMessage(
		code(1),
		"Cannot extract property from non-object"
	),
	MISCONFIGURED_DATALOADER: ErrorDescriptor.useMessageFormatter(
		code(2),
		(batchSize: number) => `Misconfigured dataloader. Missing required "maxBatchSize: ${batchSize}" option`
	),
});
