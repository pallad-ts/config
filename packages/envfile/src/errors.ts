import {Domain, generators} from 'alpha-errors';
import {ConfigError} from "@pallad/config";

export const ERRORS = new Domain({
    errorClass: ConfigError,
    codeGenerator: generators.formatCode('E_CONF_ENVFILE_%d')
}).createErrors(create => {
    return {
        ENV_FILE_DOES_NOT_EXIST: create('Env file %s does not exist but it is required'),
    };
});
