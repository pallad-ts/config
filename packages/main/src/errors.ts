import {Domain, generators} from 'alpha-errors';
import {ConfigError} from "./ConfigError";

export const ERRORS = new Domain({
    errorClass: ConfigError,
    codeGenerator: generators.formatCode('E_CONF_%d')
}).createErrors(create => {
    return {
        CONFIG_LOADING_FAILED: create('Config loading failed: \s%s'),
        PROVIDER_VALUE_NOT_AVAILABLE: create('Value not available: %s'),
        SYNC_LOADING_NOT_AVAILABLE: create('Cannot load config synchronously since one of your config value providers is asynchronous'),
        PICK_PROVIDER_UNREGISTERED_TYPE: create('Unregistered type "%s". Registered values: %s'),
        PICK_PROVIDER_TYPE_ALREADY_EXISTS: create('Type "%s" already registered'),
        CANNOT_CONVERT_TO_INT: create(`Value "%s" cannot be converted to int`),
        CANNOT_CONVERT_TO_NUMBER: create(`Value "%s" cannot be converted to number`),
        CANNOT_CONVERT_TO_BOOL: create(`Value "%s" cannot be converted to bool`),
        TYPE_URL_INVALID_PROTOCOL: create('Protocol "%s" is not allowed. Allowed: %s'),
        INVALID_PORT_NUMBER: create('Port "%s" is invalid. Allowed ranges: %s')
    };
});
