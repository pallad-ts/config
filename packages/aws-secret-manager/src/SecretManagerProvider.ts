import { left, right } from "@sweet-monads/either";
// eslint-disable-next-line @typescript-eslint/naming-convention
import DataLoader from "dataloader";
import { get } from "object-path";

import { AsyncProvider, Provider, ValueNotAvailable } from "@pallad/config";

import { SecretReference } from "./SecretReference";
import { ERRORS } from "./errors";

export class SecretManagerProvider extends AsyncProvider<unknown> {
	constructor(
		readonly reference: SecretReference,
		readonly dataLoader: DataLoader<string, unknown | undefined>
	) {
		super();
	}

	resolveValueAsync(): Promise<Provider.Value<unknown>> {
		return this.loadValue().catch(e => left(e));
	}

	private async loadValue() {
		const value = await this.dataLoader.load(this.reference.secretName);

		if (value === undefined) {
			const description = SecretReference.computeDescription(this.reference);
			return left<Provider.Fail, unknown>(new ValueNotAvailable(description));
		}

		return this.extractValue(value);
	}

	private extractValue(value: unknown) {
		if (this.reference.property) {
			// eslint-disable-next-line no-null/no-null
			if (typeof value === "object" && value !== null) {
				const resolvedValue = get(value, this.reference.property);
				// eslint-disable-next-line no-null/no-null
				if (resolvedValue === undefined || resolvedValue === null) {
					const description = SecretReference.computeDescription(this.reference);
					return left(new ValueNotAvailable(description));
				}
				return right(resolvedValue);
			}
			return left(ERRORS.CANNOT_EXTRACT_PROPERTY_FROM_NON_OBJECT.create());
		}
		return right(value);
	}
}
