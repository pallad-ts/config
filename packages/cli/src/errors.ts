import { ConfigError } from "@pallad/config";
import { Domain, ErrorDescriptor, formatCodeFactory } from "@pallad/errors";

const errorsDomain = new Domain();
const code = formatCodeFactory("E_CONF_CLI_%c");

export const ERRORS = errorsDomain.addErrorsDescriptorsMap({
    INVALID_CONFIG_VALUE_AT_PATH: ErrorDescriptor.useMessageFormatter(
        code(1),
        (parentConfigValue: unknown, propertyPath: string[]) =>
            `Cannot get config at property "${propertyPath.join(",")}" from config since parent value is of type: ${typeof parentConfigValue}`,
        ConfigError
    ),
});
