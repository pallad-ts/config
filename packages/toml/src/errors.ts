import { ConfigError } from "@pallad/config";
import { Domain, ErrorDescriptor, formatCodeFactory } from "@pallad/errors";

const errorsDomain = new Domain();
const code = formatCodeFactory("E_CONF_TOML_%c");

export const ERRORS = errorsDomain.addErrorsDescriptorsMap({
    ENV_FILE_DOES_NOT_EXIST: ErrorDescriptor.useMessageFormatter(
        code(1),
        (filePath: string) => `TOML file "${filePath}" does not exist but it is required`,
        ConfigError
    ),
});
