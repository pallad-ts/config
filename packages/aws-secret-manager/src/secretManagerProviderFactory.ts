import { SecretsManagerClient, BatchGetSecretValueCommand } from "@aws-sdk/client-secrets-manager";
import { SecretValueEntry } from "@aws-sdk/client-secrets-manager/dist-types/models/models_0";
// eslint-disable-next-line @typescript-eslint/naming-convention
import DataLoader from "dataloader";

import { SecretManagerProvider } from "./SecretManagerProvider";
import { SecretReference } from "./SecretReference";
import { ERRORS } from "./errors";

const MAX_BATCH_SIZE = 20;

function defaultCreateDataLoader(batchFN: secretManagerProviderFactory.BatchFunction) {
	return new DataLoader<string, unknown | undefined>(batchFN, {
		maxBatchSize: MAX_BATCH_SIZE,
	});
}

export function defaultDeserialize(entry: SecretValueEntry) {
	if (entry.SecretBinary) {
		return entry.SecretBinary;
	}

	if (entry.SecretString) {
		const string = entry.SecretString;
		if (string.startsWith("{") || string.startsWith("[") || string.startsWith('"')) {
			return JSON.parse(entry.SecretString);
		}

		return entry.SecretString;
	}

	throw new Error("Unsupported secret. Missing required `SecretBinary` or `SecretString` field.");
}

export function secretManagerProviderFactory({
	client = new SecretsManagerClient(),
	prefix,
	createDataLoader,
	deserialize = defaultDeserialize,
}: secretManagerProviderFactory.Options = {}) {
	const batchFn = async (keys: readonly string[]) => {
		if (keys.length > MAX_BATCH_SIZE) {
			throw ERRORS.MISCONFIGURED_DATALOADER.create(MAX_BATCH_SIZE);
		}
		const map = new Map<string, unknown>();
		const result = await client.send(
			new BatchGetSecretValueCommand({
				SecretIdList: keys.slice(),
			})
		);

		if (result.SecretValues) {
			for (const secretValue of result.SecretValues) {
				map.set(secretValue.Name!, deserialize(secretValue));
			}
		}
		return keys.map(x => map.get(x));
	};

	const dataLoader = (createDataLoader ?? defaultCreateDataLoader)(batchFn);
	return (reference: SecretReference) => {
		return new SecretManagerProvider(
			{
				name: (prefix ?? "") + reference.name,
				property: reference.property,
			},
			dataLoader
		);
	};
}

export namespace secretManagerProviderFactory {
	export interface Options {
		client?: SecretsManagerClient;
		createDataLoader?: CreateDataLoader;
		prefix?: string;
		deserialize?: (entry: SecretValueEntry) => unknown;
	}

	export type CreateDataLoader = (batchFn: BatchFunction) => DataLoader<string, unknown | undefined>;
	export type BatchFunction = (key: readonly string[]) => Promise<Array<unknown | undefined>>;
}
