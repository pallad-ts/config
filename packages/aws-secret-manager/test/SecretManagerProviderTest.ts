import { SecretsManagerClient } from "@aws-sdk/client-secrets-manager";
import { SecretManagerProvider } from "@src/SecretManagerProvider";
import { SecretReference } from "@src/SecretReference";
import { ERRORS } from "@src/errors";
import { secretManagerProviderFactory } from "@src/secretManagerProviderFactory";
import { left, right } from "@sweet-monads/either";
import * as sinon from "sinon";

import { loadAsync, ValueNotAvailable } from "@pallad/config";
import "@pallad/errors-dev";

import DataLoader = require("dataloader");

describe("SecretManagerProvider", () => {
	const REF_WITHOUT_PROPERTY = { secretName: "foo" };
	const REF_WITH_PROPERTY = { secretName: "foo", property: "bar" };

	describe("without property", () => {
		it("should return value for provided secret name", async () => {
			const batchFunction = sinon.stub().resolves(["foo"]);
			const dataLoader = new DataLoader<string, string | undefined>(batchFunction);

			const provider = new SecretManagerProvider(REF_WITHOUT_PROPERTY, dataLoader);

			await expect(provider.getValue()).resolves.toEqual(right("foo"));
			sinon.assert.calledOnceWithExactly(batchFunction, [REF_WITHOUT_PROPERTY.secretName]);
		});

		it("returns ValueNotAvailable error if secret does not exist", () => {
			const dataLoader = new DataLoader<string, string | undefined>(() => Promise.resolve([undefined]));

			const provider = new SecretManagerProvider(REF_WITHOUT_PROPERTY, dataLoader);

			return expect(provider.getValue()).resolves.toEqual(
				left(new ValueNotAvailable(SecretReference.computeDescription(REF_WITHOUT_PROPERTY)))
			);
		});
	});

	describe("with property", () => {
		it("should return value for provided secret name and property", async () => {
			const batchFunction = sinon.stub().resolves([{ bar: "foo" }]);
			const dataLoader = new DataLoader<string, Record<string, string> | undefined>(batchFunction);

			const provider = new SecretManagerProvider(REF_WITH_PROPERTY, dataLoader);

			await expect(provider.getValue()).resolves.toEqual(right("foo"));
			sinon.assert.calledOnceWithExactly(batchFunction, [REF_WITH_PROPERTY.secretName]);
		});
		it("returns ValueNotAvailable error if property does not exist", () => {
			const dataLoader = new DataLoader<string, Record<string, string> | undefined>(() => Promise.resolve([{}]));

			const provider = new SecretManagerProvider(REF_WITH_PROPERTY, dataLoader);

			return expect(provider.getValue()).resolves.toEqual(
				left(new ValueNotAvailable(SecretReference.computeDescription(REF_WITH_PROPERTY)))
			);
		});

		it("fails to retrieve value from non object", async () => {
			const dataLoader = new DataLoader<string, string | undefined>(() => Promise.resolve(["string value"]));

			const provider = new SecretManagerProvider(REF_WITH_PROPERTY, dataLoader);

			const value = await provider.getValue();

			expect(value.value).toBeErrorWithCode(ERRORS.CANNOT_EXTRACT_PROPERTY_FROM_NON_OBJECT.code);
		});
	});

	it("integration", () => {
		const factory = secretManagerProviderFactory({
			prefix: "pallad-config-",
		});

		const config = {
			fullJson: factory({ secretName: "json" }),
			fromJson: factory({ secretName: "json", property: "foo" }),
			fromJsonNested: factory({ secretName: "json", property: "nested.baz" }),
			raw: factory({ secretName: "raw" }),
		};

		const loaded = loadAsync(config);

		return expect(loaded).resolves.toMatchSnapshot();
	});
});
