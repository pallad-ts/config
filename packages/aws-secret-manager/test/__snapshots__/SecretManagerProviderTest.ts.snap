// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SecretManagerProvider integration 1`] = `
{
  "fromJson": "bar",
  "fromJsonNested": "qux",
  "fullJson": {
    "foo": "bar",
    "nested": {
      "baz": "qux",
    },
  },
  "raw": "raw_string_secret_value",
}
`;
